lib/datasets.go

```diff
@@ -74,136 +74,62 @@ type GetParams struct {
-}
+	if p.Selector == "body" {
+		if !p.All {
+			// ensure valid limit value
+			if p.Limit <= 0 {
+				p.Limit = 25
+			}
+			// ensure valid offset value
+			if p.Offset < 0 {
+				p.Offset = 0
+			}
 
-func arrayContains(subject []string, target string) bool {
-	for _, v := range subject {
-		if v == target {
-			return true
 		}
 	}
-	return false
 }
 
-// UnmarshalFromRequest implements a custom deserialization-from-HTTP request
-func (p *GetParams) UnmarshalFromRequest(r *http.Request) error {
-	mvars := mux.Vars(r)
-
-	if p == nil {
-		p = &GetParams{}
-	}
-
-	params := *p
-	if params.Ref == "" {
-		params.Ref = r.FormValue("ref")
+// Validate returns an error if GetParams fields are in an invalid state
+func (p *GetParams) Validate() error {
+	if !isValidSelector(p.Selector) {
+		return fmt.Errorf("could not parse request: invalid selector")
 	}
-
-	ref, err := dsref.Parse(params.Ref)
-	if err != nil {
-		return err
-	}
-	if ref.Username == "me" {
-		return fmt.Errorf("username \"me\" not allowed")
-	}
-
-	if sel, ok := mvars["selector"]; ok && params.Selector == "" {
-		selector, format, err := parseSelector(sel)
-		if err != nil {
-			return err
+	if p.Selector == "body" {
+		if !p.All && (p.Limit < 0 || p.Offset < 0) {
+			return fmt.Errorf("invalid limit / offset settings")
 		}
-		params.Selector = selector
-		params.Format = format
 	}
 
-	if params.Format == "" {
-		params.Format = r.FormValue("format")
-	}
+	return nil
+}
 
-	// This HTTP header sets the format to csv, and removes the json wrapper
-	if arrayContains(r.Header["Accept"], "text/csv") {
-		if params.Format != "" && params.Format != "csv" {
-			return fmt.Errorf("format %q conflicts with header \"Accept: text/csv\"", params.Format)
-		}
-		params.Format = "csv"
-		params.Selector = "body"
+func isValidSelector(selector string) bool {
+	if selector == "" {
+		return true
 	}
 
-	if params.Format != "" && params.Format != "json" && params.Format != "csv" && params.Format != "zip" {
-		return fmt.Errorf("invalid extension format")
+	match := validSelector.FindString(selector)
+	if match == "" || len(match) != len(selector) {
+		return false
 	}
-
-	// TODO(arqu): we default to true but should implement a guard and/or respect the page params
-	params.All = true
-	// listParams := ListParamsFromRequest(r)
-	// offset := listParams.Offset
-	// limit := listParams.Limit
-	// if offset == 0 && limit == -1 {
-	// 	params.All = true
-	// }
-
-	*p = params
-	return nil
+	return true
 }
 
+var validSelector = regexp.MustCompile(`^[\w-\.]*[\w]$`)
+
 // GetResult combines data with it's hashed path
 type GetResult struct {
-	Ref       *dsref.Ref       `json:"ref"`
-	Dataset   *dataset.Dataset `json:"data"`
-	Bytes     []byte           `json:"bytes"`
-	Message   string           `json:"message"`
-	FSIPath   string           `json:"fsipath"`
-	Published bool             `json:"published"`
+	Value interface{} `json:"value,omitempty"`
+	Bytes []byte      `json:"bytes,omitempty"`
```

Comment:
Is a separate `Bytes` field still needed? Only script is using it, but since `Value` is an `interface` it could just be putting that `[]byte` into `res.Value`.

Dispatch prefers method output values to be structs, in order to simplify how their allocation is done, but it's okay to just have this:
```
type GetResult struct {
  Value interface{}
}
```

Reply:
Ah, okay, so b5 and I had this discussion in chat and it never surfaced into an issue. If we don't have both fields, and value encompases all of them, this would break over rpc.`json.Marshal` would serialize the byte slice as a base 64 encoded string, but then `json.Unmarshal` would deserialize into a string. Unless it's in a field that is explicitly a byte slice, in which case `json.Unmarshal` knows to decode.

Here is part of the discussion from discord, and we decided to just go with the first option as being the least disruptive.

> Actually, right after the call I realized that .script calls will still have the issue of possibly not deserializing properly as bytes over rpc, but this can be solved by either 
> 1) keeping both GetResult.Value and GetResult.Bytes, like we have now
> 2) if we specifically don't want GetResult to only have have both fields, and instead have only one field of type interface{}, we can primarily rely on an (actual) helper function lib.GetScript that calls Get & can type assert to a byte slice OR decode from a string, as needed.
> 3) move GetScript into its own datasetMethod, which would be v clean but could be overkill
> 4) We resurrect that GetParams discussion... which does feel like overkill

Reply:
Oh right, that totally makes sense. Could you leave a comment on the field explaining this, that it's required to preserve byte slices over serialization?